File: app/src/main/java/com/icerrate/popularmovies/view/common/EndlessRecyclerOnScrollListener.java
package com.icerrate.popularmovies.view.common;

import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;

/**
 * Created by Ivan Cerrate.
 */

public abstract class EndlessRecyclerOnScrollListener extends RecyclerView.OnScrollListener {

    private boolean loading = false;
    private int visibleThreshold = 4;
    int firstVisibleItem, visibleItemCount, totalItemCount;

    private GridLayoutManager mGridLayoutManager;

    public EndlessRecyclerOnScrollListener(GridLayoutManager gridLayoutManager) {
        this.mGridLayoutManager = gridLayoutManager;
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);
        if (dy < 0) return;
        // check for scroll down
        visibleItemCount = recyclerView.getChildCount();
        totalItemCount = mGridLayoutManager.getItemCount();
        firstVisibleItem = mGridLayoutManager.findFirstVisibleItemPosition();

        if (!loading && (totalItemCount - visibleItemCount) <= (firstVisibleItem + visibleThreshold)) {
            // End has been reached, Do something
            onLoadMore();
            loading = true;
        }
    }

    public void setLoading(boolean loading) {
        this.loading = loading;
    }

    public abstract void onLoadMore();
}

File: app/src/main/java/com/icerrate/popularmovies/view/common/EndlessRecyclerOnScrollListener.kt
package com.icerrate.popularmovies.view.common

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView

/**
 * @author Ivan Cerrate.
 */
abstract class EndlessRecyclerOnScrollListener(
    private val mGridLayoutManager: GridLayoutManager
) : RecyclerView.OnScrollListener() {

    private var loading = false
    private val visibleThreshold = 0
    private var firstVisibleItem: Int = 0
    private var visibleItemCount: Int = 0
    private var totalItemCount: Int = 0

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)
        if (dy < 0) return
        // check for scroll down
        visibleItemCount = recyclerView.childCount
        totalItemCount = mGridLayoutManager.itemCount
        firstVisibleItem = mGridLayoutManager.findFirstVisibleItemPosition()

        if (!loading && (totalItemCount - visibleItemCount) <= (firstVisibleItem + visibleThreshold)) {
            // End has been reached, Do something
            onLoadMore()
            loading = true
        }
    }

    fun setLoading(loading: Boolean) {
        this.loading = loading
    }

    abstract fun onLoadMore()
}